adam_core/orbit_determination/differential_correction.py

```diff
@@ -0,0 +1,211 @@
+    mjd_tdb : float
+        Time of the state vector in MJD TDB.
+    observations : `~thor.orbit_determination.DifferentialCorrectionObservations` (N)
+        Observations to compute residuals for.
+    propagator : `~thor.propagator.Propagator`
+        Propagator to use to generate ephemeris.
+
+    Returns
+    -------
+    residuals : `~numpy.ndarray` (N,)
+        Weighted residual (chi) for each observation.
+    """
+    # Generate ephemeris and compute residuals
+    orbit = Orbits.from_kwargs(
+        coordinates=CartesianCoordinates.from_kwargs(
+            x=state_vector[0:1],
+            y=state_vector[1:2],
+            z=state_vector[2:3],
+            vx=state_vector[3:4],
+            vy=state_vector[4:5],
+            vz=state_vector[5:6],
+            time=Timestamp.from_mjd([mjd_tdb], scale="tdb"),
+            origin=Origin.from_kwargs(code=["SUN"]),
+            frame="ecliptic",
+        )
+    )
+    ephemeris = propagator.generate_ephemeris(
+        orbit, observations.observers, max_processes=1
+    )
+    residuals = Residuals.calculate(observations.coordinates, ephemeris.coordinates)
+
+    # We return the "chi2" value from adam_core.coordinates.residuals. This is equivalent to
+    # the mahalanobis distance squared (which is just the squared residual normalized by the
+    # uncertainty on the observations). This effectively allows us to the weight the
+    # observations (and their corresponding residuals) by their uncertainty.
+    return np.sqrt(residuals.chi2.to_numpy())
+
+
+def fit_least_squares(
+    orbit: Orbits,
+    observations: OrbitDeterminationObservations,
+    propagator: Propagator,
+    **kwargs
+) -> Tuple[FittedOrbits, FittedOrbitMembers]:
+    """
+    Diffentially correct a fitted orbit using least squares.
+
+    Parameters
+    ----------
+    orbit : `~adam_core.orbits.Orbits` (1)
+        Orbit to differentially correct.
+    observations : `~adam_core.orbit_determination.DifferentialCorrectionObservations` (N)
+        Observations.
+    propagator : `~adam_core.propagator.Propagator`
+        Propagator to use to generate ephemeris.
+    **kwargs
+        Additional keyword arguments to pass to `~scipy.optimize.least_squares`.
+        Some of these parameters if not specified will be set to sensible defaults.
+            xtol = 1e-12
+            ftol = 1e-12
+            gtol = 1e-12
+            jac = "2-point"
+            x_scale = "jac"
+            bounds = (-np.inf, np.inf) (for each parameter)
+
+    Returns
+    -------
+    fitted_orbit : `~adam_core.orbit_determination.FittedOrbits` (1)
+        Fitted orbit.
+    fitted_orbit_members : `~adam_core.orbit_determination.FittedOrbitMembers` (N)
+        Fitted orbit members.
+    """
+    assert len(orbit) == 1, "Only one orbit can be differentially corrected"
+
+    # TODO: Investigate whether we want to add fitting for the epoch as well
+    # Set up least squares problem
+    func = residual_function
```

Comment:
Why rename this if it's not parameterized?

Reply:
I was toying around with a second version of the function that fits the epoch as well. I'll remove this.

Reply:
I was toying around with a second version of the function that fits the epoch as well. I'll remove this.